File: Lucas.py
def calc(n, r, p):
    c = [0] * (n + 1)
    c[0] = 1
    for i in range(1, (n + 1)):
        j = min(i, r)
        while(j > 0):
            c[j] = (c[j] + c[j - 1]) % p
            j -= 1
    return c[r]
def Luc(n, r, p):
    if (r == 0):
        return 1;
    ni = int(n % p)
    ri = int(r % p)
    return (Luc(int(n / p), int(r / p), p) * calc(ni, ri, p)) % p;
n1 = input ("n: ")
r1 = input ("r: ")
p1 = input ("p: ")
n = int(n1)
r = int(r1)
p = int(p1)
print(Luc(n, r, p))

File: Lucas.java
import java.util.*;
class Lucas
{
    public static int calc(int n, int r, int p)
    {
        int c[] = new int[r+1];
        c[0] = 1;
        int m = 0;
        for(int i = 1; i <= n; i++)
        {
            m = Math.min(i, r);
            for(int j = m; j > 0; j--)
            c[j] = (c[j] + c[j-1]) % p;
        }
        return c[r];
    }
    public static int Luc(int n, int r, int p)
    {
        if(r == 0)
        return 1;
        return((Luc(n / p, r / p, p) * calc(n % p, r % p, p)) % p);
    }
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int r = sc.nextInt();
        int p = sc.nextInt();
        System.out.println(Luc(n, r, p));
    }
}